[Merged by Bors] - Implement standard keystore API

michaelsproul opened the pull request:
## Issue Addressed

Implements the standard key manager API from https://ethereum.github.io/keymanager-APIs/, formerly https://github.com/ethereum/beacon-APIs/pull/151
Related to https://github.com/sigp/lighthouse/issues/2557

## Proposed Changes

- [x] Add all of the new endpoints from the standard API: GET, POST and DELETE.
- [x] Add a `validators.enabled` column to the slashing protection database to support atomic disable + export.
- [x] Add tests for all the common sequential accesses of the API
- [x] Add tests for interactions with remote signer validators
- [x] Add end-to-end tests for migration of validators from one VC to another
- [x] Implement the authentication scheme from the standard (token bearer auth)

## Additional Info

The `enabled` column in the validators SQL database is necessary to prevent a race condition when exporting slashing protection data. Without the slashing protection database having a way of knowing that a key has been disabled, a concurrent request to sign a message could insert a new record into the database. The `delete_concurrent_with_signing` test exercises this code path, and was indeed failing before the `enabled` column was added.

The validator client authentication has been modified from basic auth to bearer auth, with basic auth preserved for backwards compatibility.

michaelsproul committed:
Implement standard keystore API

michaelsproul committed:
Add a re-registration test

michaelsproul committed:
Implement draft /auth API

michaelsproul commented:
I've added a draft implementation of the auth scheme described here: https://github.com/ethereum/beacon-APIs/pull/151#discussion_r732461095

There's a lot of discussion around the standard, so I'm going to step back from the impl until the dust settles a bit

michaelsproul committed:
Update to match latest upstream spec

michaelsproul committed:
Distinguish not_active from not_found during delete

michaelsproul committed:
More tests, fix web3signer VC API client

michaelsproul committed:
Migration tests, key cache fixes

michaelsproul commented:
This is almost ready for review!

There's one FIXME remaining about returning the correct error status for the import/delete of web3signer validators.

Other than that I'll be working on getting the API spec into the new repo: https://github.com/ethereum/keymanager-APIs, although I think we could merge _something_ to `unstable` before that is completely finalised.

We are currently divergent on the `eth/v1/auth` endpoint, so I'll also move that to a `lighthouse` prefix.

michaelsproul committed:
Stringify keystores and slashing protection

michaelsproul committed:
Fix perf of `delete_concurrent_with_signing`

michaelsproul committed:
Fix and test more web3signer interactions

michaelsproul committed:
Move auth API to /lighthouse/auth

michaelsproul committed:
Drop lockfile before delete

michaelsproul commented:
This is finally ready for review! :tada:

CI should pass now that a Windows-specific issue has been addressed in 21037cfa362270bd3ebee18274d20c2606a03766.

paulhauner commented:
I'm starting to review this! There's a conflict, FYI :slightly_smiling_face:

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into std-keymanager-api

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into std-keymanager-api

michaelsproul committed:
Don't 400 for invalid slashing protection data

michaelsproul committed:
Deny unknown fields in requests

paulhauner commented on common/eth2/src/lighthouse_vc/std_types.rs:
```diff
@@ -0,0 +1,104 @@
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
```
```suggestion
```

I'm not sure, but it seems like we don't need this.

michaelsproul replied:
Ah, this was just intended to lowercase the variants, so I've switched to `rename_all = "lowercase"` in 294d44b2b18f4ea1502851dc3f686692a07e699d

paulhauner commented on validator_client/src/http_api/keystores.rs:
```diff
@@ -0,0 +1,290 @@
+        .find(|def| def.voting_public_key == pubkey)
+    {
+        if !def.signing_definition.is_local_keystore() {
+            return Err("cannot import duplicate of existing remote signer validator".into());
```
I notice that an `enabled = false` validator might not appear in `/list` but they'll trigger this duplicate warning here. I don't see this as a significant issue and I assume there's going to be some edge-cases when meshing the standard API with our validator management.

michaelsproul replied:
Yeah good point. I'm OK with the current behaviour, but yeah, might make sense to re-assess in future.

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -456,17 +572,24 @@ impl InitializedValidators {
+        // Read relevant key stores from the filesystem.
         let mut definitions_map = HashMap::new();
-        for def in self.definitions.as_slice() {
+        for def in self.definitions.as_slice().iter().filter(|def| def.enabled) {
```
Hmm, it's interesting that we didn't previously filter `enabled = false` :thinking: I think it's definitely the right thing to do.

paulhauner reviewed (approved):
This is great, really neat and well tested :ok_hand: I don't have anything substantial to say, I just had a few comments whilst reading it.

It's great to see the standard API progressing :tada:

michaelsproul committed:
Clarify `ImportKeystoreStatus` attribute

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into std-keymanager-api

michaelsproul commented:
Thanks for the review @paulhauner! Will be great to get this out into a release

I think I've addressed the review comments so this should be good to go :crossed_fingers:

paulhauner reviewed (approved):
# :rocket:

michaelsproul commented:
bors r+
